packages/parser/src/envelopes/snssqs.ts

```diff
@@ -0,0 +1,185 @@
+import { ZodError, type ZodIssue, type ZodSchema, type z } from 'zod';
+import { ParseError } from '../errors.js';
+import { SnsSqsNotificationSchema } from '../schemas/sns.js';
+import { SqsSchema } from '../schemas/sqs.js';
+import type { ParsedResult, SnsSqsNotification } from '../types/index.js';
+import { envelopeDiscriminator } from './envelope.js';
+
+const createError = (index: number, issues: ZodIssue[]) => ({
+  issues: issues.map((issue) => ({
+    ...issue,
+    path: ['Records', index, 'body', ...issue.path],
+  })),
+});
+
+type ParseStepSuccess<T> = {
+  success: true;
+  data: T;
+};
+
+type ParseStepError = {
+  success: false;
+  error: { issues: ZodIssue[] };
+};
+
+type ParseStepResult<T> = ParseStepSuccess<T> | ParseStepError;
+
+const parseStep = <U>(
+  parser: (data: unknown) => z.SafeParseReturnType<unknown, U>,
+  data: unknown,
+  index: number
+): ParseStepResult<U> => {
+  const result = parser(data);
+  return result.success
+    ? { success: true, data: result.data }
+    : {
+        success: false,
+        error: createError(index, result.error.issues),
+      };
+};
+
+/**
+ * SNS plus SQS Envelope to extract array of Records
+ *
+ * Published messages from SNS to SQS has a slightly different payload structure
+ * than regular SNS messages, and when sent to SQS, they are stringified into the
+ * `body` field of each SQS record.
+ *
+ * To parse the `Message` field of the SNS notification, we need to:
+ * 1. Parse SQS schema with incoming data
+ * 2. `JSON.parse()` the SNS payload and parse against SNS Notification schema
+ * 3. Finally, parse the payload against the provided schema
+ */
+export const SnsSqsEnvelope = {
+  /**
+   * This is a discriminator to differentiate whether an envelope returns an array or an object
+   * @hidden
+   */
+  [envelopeDiscriminator]: 'array' as const,
+  parse<T extends ZodSchema>(data: unknown, schema: T): z.infer<T>[] {
+    let parsedEnvelope: z.infer<typeof SqsSchema>;
+    try {
+      parsedEnvelope = SqsSchema.parse(data);
+    } catch (error) {
+      throw new ParseError('Failed to parse SQS Envelope', {
+        cause: error as Error,
+      });
+    }
+
+    return parsedEnvelope.Records.map((record, recordIndex) => {
+      try {
+        return schema.parse(
+          SnsSqsNotificationSchema.parse(JSON.parse(record.body)).Message
+        );
+      } catch (error) {
+        throw new ParseError(
+          `Failed to parse SQS Record at index ${recordIndex}`,
+          {
+            cause: new ZodError(
+              error instanceof ZodError
+                ? (error as ZodError).issues.map((issue) => ({
+                    ...issue,
+                    path: ['Records', recordIndex, 'body', ...issue.path],
```

Comment:
Now that we have `createError` helper, do you think it'd make sense to use it here, and add later to other envelopes? We probably need ot pass the 2nd parameter as input, as it's often different. What do you think?

Reply:
Yea, possibly.

I think let's go through all the envelopes and then we see where we can simplify / extract further. Using this function removes also the `// @ts-expect-error` since it's now properly typed.

I'd rather complete the entire refactor first though, as I don't want to backtrack before getting to the end. Ideally I'd like to make it to the next release with most/all the envelopes bugs fixed - then we can refactor and optimize.